src/solidity-frontend/solidity_convert.cpp

```diff
@@ -0,0 +1,1898 @@
+{
+  unsigned index = 0;
+  nlohmann::json ast_nodes = contract_def["nodes"];
+  for(nlohmann::json::iterator itr = ast_nodes.begin(); itr != ast_nodes.end();
+      ++itr, ++index)
+  {
+    nlohmann::json ast_node = *itr;
+    std::string node_name = ast_node["name"].get<std::string>();
+    std::string node_type = ast_node["nodeType"].get<std::string>();
+    msg.status(fmt::format(
+      "@@ Converting node[{}]: name={}, nodeType={} ...",
+      index,
+      node_name.c_str(),
+      node_type.c_str()));
+    exprt dummy_decl;
+    if(get_decl(ast_node, dummy_decl))
+      return true;
+  }
+
+  assert(current_functionDecl == nullptr);
+
+  return false;
+}
+
+bool solidity_convertert::get_decl(
+  const nlohmann::json &ast_node,
+  exprt &new_expr)
+{
+  new_expr = code_skipt();
+
+  if(!ast_node.contains("nodeType"))
+    assert(!"Missing \'nodeType\' filed in ast_node");
+
+  SolidityGrammar::ContractBodyElementT type =
+    SolidityGrammar::get_contract_body_element_t(ast_node);
+
+  // based on each element as in Solidty grammar "rule contract-body-element"
+  switch(type)
+  {
+  case SolidityGrammar::ContractBodyElementT::StateVarDecl:
+  {
+    return get_var_decl(ast_node, new_expr); // rule state-variable-declaration
+  }
+  case SolidityGrammar::ContractBodyElementT::FunctionDef:
+  {
+    return get_function_definition(ast_node); // rule function-definition
+  }
+  default:
+  {
+    assert(!"Unimplemented type in rule contract-body-element");
+    return true;
+  }
+  }
+
+  return false;
+}
+
+bool solidity_convertert::get_var_decl_stmt(
+  const nlohmann::json &ast_node,
+  exprt &new_expr)
+{
+  // For rule variable-declaration-statement
+  new_expr = code_skipt();
+
+  if(!ast_node.contains("nodeType"))
+    assert(!"Missing \'nodeType\' filed in ast_node");
+
+  SolidityGrammar::VarDeclStmtT type =
+    SolidityGrammar::get_var_decl_stmt_t(ast_node);
+  msg.status(fmt::format(
+    "	@@@ got Variable-declaration-statement: "
+    "SolidityGrammar::VarDeclStmtT::{}",
+    SolidityGrammar::var_decl_statement_to_str(type)));
+
+  switch(type)
+  {
+  case SolidityGrammar::VarDeclStmtT::VariableDecl:
+  {
+    return get_var_decl(ast_node, new_expr); // rule variable-declaration
+  }
+  default:
+  {
+    assert(!"Unimplemented type in rule variable-declaration-statement");
+    return true;
+  }
+  }
+
+  return false;
+}
+
+bool solidity_convertert::get_var_decl(
+  const nlohmann::json &ast_node,
+  exprt &new_expr)
+{
+  // For Solidity rule state-variable-declaration:
+  // 1. populate typet
+  typet t;
+  // VariableDeclaration node contains both "typeName" and "typeDescriptions".
+  // However, ExpressionStatement node just contains "typeDescriptions".
+  // For consistensy, we use ["typeName"]["typeDescriptions"] as in state-variable-declaration
+  // to improve the re-usability of get_type* function, when dealing with non-array var decls.
+  // For array, do NOT use ["typeName"]. Otherwise, it will cause problem
```

Comment:
> Do you mean in this function or during the parsing phase?

It isn't about the parsing phase. Everything in `solidity_convert.cpp` is associated with typechecking phase.

> If it is during this function, is there a way to convert this comment into an assertion?

I don't think it can be converted into an assertion. Because the whole purpose of these comments is to remind people why using ["typeName"]["typeDescription"] instead of other sources type information in Solidity JSON AST. By "other sources", I mean: 
![image](https://user-images.githubusercontent.com/32592800/137314736-4221fe52-fab6-4a33-b896-f11a66ac81b5.png)

I believe comments are good enough here.